src/Application/Implementation/StatePool.cs

```diff
@@ -30,9 +30,17 @@ public void Add(IExecutable executable)
             {
                 try
                 {
-                    foreach (var child in executable.Run())
-                        if (_exception == null)
-                            Add(child);
+                    foreach (var forks in executable.Run())
+                    {
+                        if (_exception != null)
+                        {
+                            break;
+                        }
+                        foreach (var fork in forks)
```

Comment:
Writing it this way has made it a little more general purpose than we perhaps need. Specifically, this code would now support a scenario where the `State` could generate forks but that not be terminal. As in, it could fork but keep going itself. The way it is modelled is basically as a stream of forks, which  right now means a stream of empty lists until the very final one which may contain two new states if this state does indeed fork.

Reply:
In case you want to see the upshot of this in terms of them collecting stats, I've added a commit to the `executable-stats` branch so PR #90 shows what that looks like under this regime too.